File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-csharp/src/SDKWork.Mail.Sdk/mailProviderPackageCatalog.cs
namespace Sdkwork.Mail.Sdk;

using System.Linq;

public sealed record MailProviderPackageCatalogEntry(
    string providerKey,
    string pluginId,
    string driverId,
    string packageIdentity,
    string manifestPath,
    string readmePath,
    string sourcePath,
    string sourceSymbol,
    bool builtin,
    bool rootPublic,
    string status,
    string runtimeBridgeStatus
);

public static class MailProviderPackageCatalog
{
    public static readonly IReadOnlyList<MailProviderPackageCatalogEntry> Entries =
    [
        new("smtp", "Mail-smtp", "sdkwork-mail-driver-smtp", "Sdkwork.Mail.Sdk.Provider.Smtp", "providers/Sdkwork.Mail.Sdk.Provider.Smtp/Sdkwork.Mail.Sdk.Provider.Smtp.csproj", "providers/Sdkwork.Mail.Sdk.Provider.Smtp/README.md", "providers/Sdkwork.Mail.Sdk.Provider.Smtp/src/MailProviderSmtpPackageContract.cs", "MailProviderSmtpPackageContract", true, false, "future-runtime-bridge-only", "reserved"),
        new("imap", "Mail-imap", "sdkwork-mail-driver-imap", "Sdkwork.Mail.Sdk.Provider.Imap", "providers/Sdkwork.Mail.Sdk.Provider.Imap/Sdkwork.Mail.Sdk.Provider.Imap.csproj", "providers/Sdkwork.Mail.Sdk.Provider.Imap/README.md", "providers/Sdkwork.Mail.Sdk.Provider.Imap/src/MailProviderImapPackageContract.cs", "MailProviderImapPackageContract", true, false, "future-runtime-bridge-only", "reserved"),
    ];

public static MailProviderPackageCatalogEntry? GetMailProviderPackageByProviderKey(string providerKey) =>
        Entries.FirstOrDefault(entry => entry.providerKey == providerKey);

    public static MailProviderPackageCatalogEntry? GetMailProviderPackageByPackageIdentity(string packageIdentity) =>
        Entries.FirstOrDefault(entry => entry.packageIdentity == packageIdentity);

}

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-java/src/main/java/com/sdkwork/Mail/metadata/mailProviderPackageCatalog.java
package com.sdkwork.Mail.metadata;

import java.util.List;
import java.util.Optional;

public final class MailProviderPackageCatalog {

  public static final List<MailProviderPackageCatalogEntry> ENTRIES = List.of(
      new MailProviderPackageCatalogEntry("smtp", "Mail-smtp", "sdkwork-mail-driver-smtp", "com.sdkwork:Mail-sdk-provider-smtp", "providers/Mail-sdk-provider-smtp/pom.xml", "providers/Mail-sdk-provider-smtp/README.md", "providers/Mail-sdk-provider-smtp/src/main/java/com/sdkwork/Mail/provider/smtp/MailProviderSmtpPackageContract.java", "MailProviderSmtpPackageContract", true, false, "future-runtime-bridge-only", "reserved"),
      new MailProviderPackageCatalogEntry("imap", "Mail-imap", "sdkwork-mail-driver-imap", "com.sdkwork:Mail-sdk-provider-imap", "providers/Mail-sdk-provider-imap/pom.xml", "providers/Mail-sdk-provider-imap/README.md", "providers/Mail-sdk-provider-imap/src/main/java/com/sdkwork/Mail/provider/imap/MailProviderImapPackageContract.java", "MailProviderImapPackageContract", true, false, "future-runtime-bridge-only", "reserved")
  );

public static Optional<MailProviderPackageCatalogEntry> getMailProviderPackageByProviderKey(String providerKey) {
    for (var entry : ENTRIES) {
      if (entry.providerKey().equals(providerKey)) {
        return Optional.of(entry);
      }
    }

    return Optional.empty();
  }

  public static Optional<MailProviderPackageCatalogEntry> getMailProviderPackageByPackageIdentity(String packageIdentity) {
    for (var entry : ENTRIES) {
      if (entry.packageIdentity().equals(packageIdentity)) {
        return Optional.of(entry);
      }
    }

    return Optional.empty();
  }


  private MailProviderPackageCatalog() {
  }

  public record MailProviderPackageCatalogEntry(
      String providerKey,
      String pluginId,
      String driverId,
      String packageIdentity,
      String manifestPath,
      String readmePath,
      String sourcePath,
      String sourceSymbol,
      boolean builtin,
      boolean rootPublic,
      String status,
      String runtimeBridgeStatus
  ) {
  }
}
